### ISSUE TITLE
Expandable Objects

### ISSUE THREAD

**Issue by Buratti** [vapor-community/stripe-kit#15]
👋 

I just push a very basic draft about how we could implement expandable objects.
It is a braking change so before going on I'd like to have a feedback.
You can find the diff [here](https://github.com/vapor-community/StripeKit/compare/master...Buratti:master).

The result would be
```swift
let customer = try stripe.customers.retrieve(customer: "cus_123", expand: ["default_source"]).wait()

print(customer.defaultSource?.last4) // 4242
print(customer.$defaultSource.id) //card_1234
```
and
```swift
let customer = try stripe.customers.retrieve(customer: "cus_123").wait()

print(customer.defaultSource?.last4) // nil
print(customer.$defaultSource.id) //card_1234
```

Cheers 🍻

**Comment by Andrewangeta** [comment]
@Buratti I love the idea of using property wrappers. And you're in the same vein of how I was thinking of implementing it with a `StripeExpandable` protocol etc. We can dive a bit more into it when you get a chance.

**Comment by Andrewangeta** [comment]
@anthonycastelli @ksmandersen thoughts on this approach?

**Comment by anthonycastelli** [comment]
I’m all for it. Right now Stripe and Vapor still aren’t at a full official release so a breaking change isn’t a big deal right now. I like the idea of incorporating property wrappers this way too. My only question is what is the benefit of doing this?

**Comment by ksmandersen** [comment]
I like this approach. 

My main concern with expandable objects is the difference in types. E.g if you don’t expand the `latestInvoice` on `Subscription` it is `String?` if you do expand it becomes a `Invoice?`. The key in the json is the same.

**Comment by Andrewangeta** [comment]
@ksmandersen Yea so when you define a property or mark it with `@Expandable` you'd specify the type as the `String?` and from there if you wanted to use the projectedValue or the expandable type, you'd use the `$` syntax like 
`customer.$defaultSource(as: StripeCard.self)?.last4`
`customer.$defaultSource(as: StripeBankAccount.self).?last4`

And you'd still get the id as normal 
`customer.defaultSource // src_1234`

**Comment by Buratti** [comment]
@ksmandersen that's one of the reason beyond property wrappers. If you decide to not expand latestInvoice you are still able to access its id property as a string using the projected value (`$latestInvoice.id`).

There might be a smarter solution though. A couple of days ago Andrew and I were discussing on Discord about the "multi type" problem. The default source of a user can be a Card object but it might also be another type of payment method. The current approach would not be able to handle this situation.

A different solution might be to have
```swift
@Expandable
var defaultSource: String?
```

If we always keep `String` as the wrapped value we can project some other object. The projected object will be responsible to decode the expanded object as the user request.
For example:
```swift
customer.defaultSource // "card_1234" <- not source breaking, the wrapped value will always be the object id or nil if no defaultSource exists.
customer.$defaultSource.decoded(as: StripeCard.self)?.last4 // "4242" if the expanded object was a Card, nil if it was a different form of Payment Method.

```

This solution is a bit uglier from a syntax point of view but with Swift 5.2 we can improve from 
```swift
customer.$defaultSource.decoded(as: StripeCard.self)?.last4

````
to
```swift
customer.$defaultSource(as: StripeCard.self)?.last4
```

which is not so bad.

**Comment by Buratti** [comment]
Regarding the same key problem, you can find a custom implementation of `init(from decoder: Decoder) throws` [here](https://github.com/vapor-community/StripeKit/compare/master...Buratti:master)
If the object is not expanded, the key for `latestInvoice` will map to a `single value container` which will contain a string. If the object is expanded, the key will map to a `keyed container` keyed by `StripeCard.CodingKeys.self`.
Therefor, if the object is expanded `String.init(from: decoder)` will fail and if the object is not expanded `StripeCard.init(from: decoder)` will fail.

**Comment by Andrewangeta** [comment]
This was implemented in #72
